### PR TITLE
ramips: convert MT7621 wireless calibration EEPROM to NVMEM format 

### PR DESCRIPTION
**Pull request by DragonBluep**
Notice:
1. Some MT7621 devices have complex dtsi dependencies, so they are not included.
2. All NIC models are judged based on the commit message or the makefile default packages.

|Package Name|NIC Model|EEPROM size||
|-|-|-|-|
|kmod-mt7603|MT7603|0x400|2 GHz only|
|kmod-mt76x0e|MT7610|0x200|same as MT7612|
|kmod-mt76x2|MT7602/MT7612|0x200||
|kmod-mt7663-firmware-ap|MT7613|0x4da8|same as MT7615|
|kmod-mt7615-firmware|MT7615|0x4da8||
|kmod-mt7915-firmware|MT7915|0xe00||

### CODE REVIEW

File: target/linux/ramips/dts/mt7621_linksys_e7350.dts [hunk 1]
```diff
@@ -127,7 +134,8 @@
 	wifi@0,0 {
```
Comment (neheb) [inline target/linux/ramips/dts/mt7621_linksys_e7350.dts:134]: now I'm curious what happens if this section is removed. Does WiFi go away?

File: target/linux/ramips/dts/mt7621_snr_snr-cpe-me1.dts [hunk 2]
```diff
@@ -144,13 +145,20 @@
 			};
 
 			factory: partition@40000 {
+				compatible = "nvmem-cells";
 				label = "Factory";
 				reg = <0x40000 0x10000>;
-				read-only;
```
Comment (csharper2005) [inline target/linux/ramips/dts/mt7621_snr_snr-cpe-me1.dts:149]: Why are we remove `read-only`?

**Comment by raenye** [comment]
Does NVMEM format make it easier to patch EEPROM via DTS?
(instead of going through the `insmod mtr_rw i-want-a-brick` route)

**Comment by DragonBluep** [comment]
> Does NVMEM format make it easier to patch EEPROM via DTS? (instead of going through the `insmod mtr_rw i-want-a-brick` route)

No, NVMEM means `Non Volatile Memory`. It can't be edited.

**Comment by dangowrt** [comment]
> No, NVMEM means `Non Volatile Memory`. It can't be edited.

Sorry, but that's not entirely correct. Quoting Wikipedia: "Non-volatile random-access memory (NVRAM) is random-access memory that retains data without applied power." It can totally offer read-write access.

Using the nvmem framework in Linux on top of an MTD device results in read-only access simply because write access isn't implemented in the NVMEM MTD backend driver (because one can just use MTD API to write, for other NVMEM devices, e.g. AT24/AT25 I2C EEPROMs, write via NVMEM API *is* implemented).

tldr; This is unrelated. You will still need `mtd-rw` to remove the read-only flag from MTD partitions if you wish to modify their content. Nothing to do with whether the driver uses nvmem-cells or our previous mtd-eeprom hack to access the WiFi EEPROM data.

**Review by Ansuel — CHANGES_REQUESTED** [review]
Man I think after this review, I'm ready for a run of Paper Please...

Generally this looks ok. I notice there are additional cleanup but they are ok and can stay considering the amount of work done. Just some minor tweak to do.

Also please do them but don't rebase this on top of main so I can check the changes by comparing the commit before and after. (if you are asking yes I checked all the 148 file)

File: target/linux/ramips/dts/mt7621_tplink_ec330-g5u-v1.dts [hunk 3]
```diff
@@ -225,25 +225,30 @@
...
 				#address-cells = <1>;
 				#size-cells = <1>;
 
+				eeprom_factory_8000: eeprom@8000 {
+					reg = <0x8000 0x4da8>;
+				};
+
+				eeprom_factory_14000: eeprom@14000 {
+					reg = <0x14000 0x4da8>;
+				};
```
[the remark this answers is not in the document]
Comment (DragonBluep) [reply target/linux/ramips/dts/mt7621_tplink_ec330-g5u-v1.dts:244]: It should be okay. I saw @rmilecki pushed a similar commit. https://github.com/openwrt/openwrt/commit/03b328480559a0183752f668dece2ff00d3bee28

**Comment by DragonBluep** [comment]
@Ansuel Hi! Thanks for your review. I've made the changes according to your suggestion. And when I checked if nesting eeprom cell into nvmem-layout was the correct way, I found the EEPROM size of the I-O DATA WN-DEAX1800GR was wrong, so I added another small commit to fix it. https://github.com/openwrt/openwrt/pull/13587/commits/6b74f1b708ed6c5f1709312300c51d854e94cf71

File: target/linux/ramips/dts/mt7621_bolt_arion.dts [hunk 4]
```diff
@@ -100,9 +100,24 @@
...
+				eeprom_factory_0: eeprom@0 {
+					reg = <0x0 0x400>;
+				};
+
+				eeprom_factory_8000: eeprom@8000 {
+					reg = <0x8000 0x200>;
+				};
+
+				macaddr_factory_4000: macaddr@4000 {
+					reg = <0x28 0x6>;
```
[the remark this answers is not in the document]
Comment (Ansuel) [reply target/linux/ramips/dts/mt7621_bolt_arion.dts]: Idea for this fix was to change this in a previous commit and then make the convert commit just move the node inside.
Since you have to touch this pr again, can you do this change?

[the remark this answers is not in the document]
Comment (DragonBluep) [reply target/linux/ramips/dts/mt7621_bolt_arion.dts]: You mean nesting nvmem stuff into partition nodes in the previous commit? It should be fine now.
https://github.com/openwrt/openwrt/pull/13587/commits/6f52c1429a29e0594df10d7df223366a41e29ff9

[the remark this answers is not in the document]
Comment (Ansuel) [reply target/linux/ramips/dts/mt7621_bolt_arion.dts]: No I mean that before all these commit, you create a single simple commit that just fix the wrong node name and tag.

And then in the conversion you move the node in the partition nodes as you did in the linked commit.

[the remark this answers is not in the document]
Comment (DragonBluep) [reply target/linux/ramips/dts/mt7621_bolt_arion.dts]: Oh, it seems that you didn't notice this small commit before, I've already done it. https://github.com/openwrt/openwrt/commit/001f05b220e3d2a0419026ce50332c8b9a33cc69
I'll force push it. Thanks again. It's really a hard job to review these changes.

[the remark this answers is not in the document]
Comment (Ansuel) [reply target/linux/ramips/dts/mt7621_bolt_arion.dts]: Oh it was there and I didn't notice! Sorry!!

**Comment by raenye** [comment]
Please see issues #13700 and #13701 (mt7915).

**Comment by neheb** [comment]
this is not in 23.05.0

**Comment by raenye** [comment]
@neheb #13701 is indeed mt7622 so probably not related, but #13700 is snapshot.
